File: kraken/pytorch/initializer.py
# coding=utf-8

from typing import Dict
import kraken_native


class Initializer(object):

  def type(self) -> kraken_native.InitializerType:
    raise NotImplementedError

  def conf(self) -> Dict[str, str]:
    raise NotImplementedError


class ConstantInitializer(Initializer):

  def __init__(self, value: float = 0.0):
    super(ConstantInitializer, self).__init__()
    self._value = value

  def type(self) -> kraken_native.InitializerType:
    return kraken_native.InitializerType.kConstant

  def conf(self) -> Dict[str, str]:
    return {'value': str(self._value)}


class UniformInitializer(Initializer):

  def __init__(self, lower: float = 0.0, upper: float = 1.0):
    super(UniformInitializer, self).__init__()
    self._lower = lower
    self._upper = upper

  def type(self) -> kraken_native.InitializerType:
    return kraken_native.InitializerType.kUniform

  def conf(self) -> Dict[str, str]:
    return {'lower': str(self._lower), 'upper': str(self._upper)}


class NormalInitializer(Initializer):

  def __init__(self, mean: float = 0.0, stddev: float = 1.0):
    super(NormalInitializer, self).__init__()
    self._mean = mean
    self._stddev = stddev

  def type(self) -> kraken_native.InitializerType:
    return kraken_native.InitializerType.kNormal

  def conf(self) -> Dict[str, str]:
    return {'mean': str(self._mean), 'stddev': str(self._stddev)}


class XavierUniformInitializer(Initializer):

  def __init__(self, gain: float = 1.0):
    self._gain = gain

  def type(self) -> kraken_native.InitializerType:
    return kraken_native.InitializerType.kXavierUniform

  def conf(self) -> Dict[str, str]:
    return {'gain': str(self._gain)}


class XavierNormalInitializer(Initializer):

  def __init__(self, gain: float = 1.0):
    self._gain = gain

  def type(self) -> kraken_native.InitializerType:
    return kraken_native.InitializerType.kXavierNormal

  def conf(self) -> Dict[str, str]:
    return {'gain': str(self._gain)}

File: kraken/ps/initializer/initializer.cc
#include "ps/initializer/initializer.h"

#include "ps/initializer/constant_initializer.h"
#include "ps/initializer/normal_initializer.h"
#include "ps/initializer/uniform_initializer.h"
#include "ps/initializer/xavier_normal_initializer.h"
#include "ps/initializer/xavier_uniform_initializer.h"

namespace kraken {

Initializer::Initializer(InitializerType type) : type_(type) {
}

InitializerType Initializer::type() const {
  return type_;
}

std::unordered_map<std::string, std::string> Initializer::conf() const {
  return {};
}

std::unique_ptr<Initializer> Initializer::Create(
    InitializerType init_type,
    const std::unordered_map<std::string, std::string>& init_conf) {
  std::unique_ptr<Initializer> initializer;

  if (init_type == InitializerType::kConstant) {
    float value = 0;
    utils::ParseConf<float>(init_conf, "value", &value);

    initializer.reset(new ConstantInitializer(value));
  } else if (init_type == InitializerType::kNormal) {
    float mean = 0.0;
    float stddev = 1.0;

    utils::ParseConf<float>(init_conf, "mean", &mean);
    utils::ParseConf<float>(init_conf, "stddev", &stddev);

    initializer.reset(new NormalInitializer(mean, stddev));
  } else if (init_type == InitializerType::kUniform) {
    float lower = 0.0;
    float upper = 1.0;

    utils::ParseConf<float>(init_conf, "lower", &lower);
    utils::ParseConf<float>(init_conf, "upper", &upper);

    initializer.reset(new UniformInitializer(lower, upper));
  } else if (init_type == InitializerType::kXavierNormal) {
    float gain = 1.0;
    utils::ParseConf<float>(init_conf, "gain", &gain);

    initializer.reset(new XavierNormalInitializer(gain));
  } else if (init_type == InitializerType::kXavierUniform) {
    float gain = 1.0;
    utils::ParseConf<float>(init_conf, "gain", &gain);

    initializer.reset(new XavierUniformInitializer(gain));
  }

  return initializer;
}

}  // namespace kraken
